config/options.go

```diff
@@ -499,21 +530,59 @@ func (o *Options) Validate() error {
 		o.Headers = make(map[string]string)
 	}
 
-	if o.InsecureServer {
-		log.Warn().Msg("config: insecure mode enabled")
-	} else if o.Cert != "" || o.Key != "" {
-		o.TLSCertificate, err = cryptutil.CertifcateFromBase64(o.Cert, o.Key)
-	} else if o.CertFile != "" || o.KeyFile != "" {
-		o.TLSCertificate, err = cryptutil.CertificateFromFile(o.CertFile, o.KeyFile)
-	} else {
-		err = errors.New("config:no certificates supplied nor was insecure mode set")
+	if o.Cert != "" || o.Key != "" {
+		o.TLSConfig, err = cryptutil.TLSConfigFromBase64(o.TLSConfig, o.Cert, o.Key)
+		if err != nil {
+			return fmt.Errorf("config: bad cert base64 %w", err)
+		}
 	}
-	if err != nil {
-		return err
+
+	if len(o.Certificates) != 0 {
+		for _, c := range o.Certificates {
+			o.TLSConfig, err = cryptutil.TLSConfigFromFile(o.TLSConfig, c.CertFile, c.KeyFile)
+			if err != nil {
+				return fmt.Errorf("config: bad cert file %w", err)
+			}
+		}
+	}
+
+	if o.CertFile != "" || o.KeyFile != "" {
+		o.TLSConfig, err = cryptutil.TLSConfigFromFile(o.TLSConfig, o.CertFile, o.KeyFile)
+		if err != nil {
+			return fmt.Errorf("config: bad cert file %w", err)
+		}
+	}
+	if o.AutoCert {
+		o.TLSConfig, o.AutoCertHandler, err = cryptutil.NewAutocert(
+			o.TLSConfig,
+			o.sourceHostnames(),
+			o.AutoCertUseStaging,
+			o.AutoCertFolder)
+		if err != nil {
+			return fmt.Errorf("config: autocert failed %w", err)
+		}
+	}
+	if !o.InsecureServer && o.TLSConfig == nil {
+		return fmt.Errorf("config: server must be run with `autocert`, " +
+			"`insecure_server` or manually provided certificates to start")
 	}
 	return nil
 }
 
+func (o *Options) sourceHostnames() []string {
+	if len(o.Policies) == 0 {
+		return nil
+	}
+	var h []string
+	for _, p := range o.Policies {
+		h = append(h, p.Source.Hostname())
+	}
+	if o.AuthenticateURL != nil {
```

Comment:
do we need to add forward auth too?

Reply:
yes, good catch!

Reply:
actually, on second thought, I think this is best left as an exercise to the user since in any forward auth configuration I'd expect the operator would have some other system managing public certificates like cert-manager, and terminating TLS at that fronting proxy. Thoughts @calebdoxsey ?

Reply:
:+1: